### PR TITLE
Spotify scraper and HTML rendering prototype

### PR DESCRIPTION
**Pull request by addiebarron**
## Description

Created a content script in `source/content/index.js` that scrapes spotify web, passes the result to [bandcamp search](github.com/addiebarron/bandcamp-search-scraper) and implements basic HTML rendering for the returned bandcamp data. Had some fun with CSS too.

Since spotify web is a single-page app, a background script `source/background.js` listens for URL changes and notifies our content script on these "page changes". The content script then uses [arrive](https://github.com/uzairfarooq/arrive) to watch for the creation of elements matching the selectors of interest.

Added Babel transpiling via webpack to support async/await and some useful futuristic features like optional chaining.

## Browsers
Only tested on a Firefox dev instance, but theoretically this is all cross platform code :)

## Testing
Take it for a spin with `npm run dev` and visit an album/artist page! Content script errors are logged in the page's console, while background script / extension errors are logged in about:debugging > This Firefox > Inspect.

<3

### CODE REVIEW

**Comment by jeancochrane** [comment]
So excited to take this for a whirl!

**Comment by addiebarron** [comment]
Yay! Btw, realizing I branched from my bandcamp-scraper branch rather than master so the `options` proof-of-concept code is in here too. Feel free to ignore that, we can write an options page later if we want.

File: source/content/index.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  // returns a bandcamp search result object or null
+}
+
+// get artist name from the current spotify SPA view
+function spotifyScrape() {
+  return new Promise((resolve) => {
+    document.unbindArrive();
+    for (let type in selectors) {
+      document.arrive(
+        selectors[type].artist,
```
Comment (jeancochrane) [inline source/content/index.js:79]: This totally works, but you could also use [`Object.entries()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) to iterate the object!

```suggestion
    for (let [type, selector] in selectors) {
      document.arrive(
        selector.artist,
```

File: source/content/index.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+// establish a port for communicating with the background script
+let port = browser.runtime.connect({ name: 'url-change-port' });
+// run our code each time the SPA url changes
+port.onMessage.addListener(onURLChange);
+// run once on page load
+onURLChange({ url: window.location.href });
+
+// to be run when the page URL changes. can't be async, for webextension reasons:
+// https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#Parameters
```
Comment (jeancochrane) [inline source/content/index.js:16]: Good comment here 👍🏻

File: source/options/index.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// import optionsStorage from './options-storage';
+
+// optionsStorage.syncForm('#search-form');
+
```
Comment (jeancochrane) [inline source/options/index.js:4]: Things we can get rid of maybe?

```suggestion
```

File: source/background.js [hunk 4]
```diff
@@ -1,2 +1,26 @@
 // eslint-disable-next-line import/no-unassigned-import
-import './options-storage';
+// import './options/options-storage';
+
```
Comment (jeancochrane) [inline source/background.js:3]: This looks like it's not used, is that right?

```suggestion
```

File: source/options/options.html [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+		<span>Search</span>
+		<input type="text" placeholder="Search on Bandcamp">
+	</label>
+</form>
+
+<pre id="results-pre">
+Results will appear here.
+</pre>
+
+<!-- <script src="browser-polyfill.min.js"></script> -->
```
Comment (jeancochrane) [inline source/options/options.html:18]: Another stray comment:

```suggestion
```

File: source/options/options-storage.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+// 	defaults: {
+// 		colorRed: 244,
+// 		colorGreen: 67,
+// 		colorBlue: 54
+// 	},
+// 	migrations: [
+// 		OptionsSync.migrations.removeUnused
+// 	],
+// 	logging: true
+// });
```
Comment (jeancochrane) [inline source/options/options-storage.js:13]: We can probably get rid of this whole file, right?

File: source/content/index.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+  // remove loader
+  $bcNudge.find('.loader').remove();
+
+  // render final state
+  if (bcResult) {
+    $bcNudge
+      .addClass('success')
+      .append(
+        `<a target="_blank" href="${bcResult.url}">Found a match on Bandcamp</a>`
```
Comment (jeancochrane) [inline source/content/index.js:49]: Maybe something a bit more forward here?

```suggestion
        `<a target="_blank" href="${bcResult.url}">Support this artist on Bandcamp</a>`
```

**Review by jeancochrane — APPROVED** [review]
I love this! It works really well and looks beautiful 😊 A couple of nitpicks below but they're extremely minor.

The `options` module is providing content for the plugin popup, right? It seems like it's useful for debugging for now but in the future we might want to think about getting rid of it or presenting it in a way that makes it clearer what it's for. But that's a decision for a future time!

File: source/content/index.js (hunk 1 above)
Comment (addiebarron) [reply source/content/index.js:79]: I like how that looks better! Never really understood why for/in loops can't iterate over keys & values at the same time.

**Pull request merged**